crates/runtime/src/actor/mod.rs

```diff
@@ -0,0 +1,438 @@
+    ///
+    /// Fails if [`Instance::call`] fails
+    #[instrument(skip_all)]
+    pub async fn call(
+        &self,
+        operation: impl AsRef<str>,
+        request: impl AsyncRead + Send + Sync + Unpin + 'static,
+        response: impl AsyncWrite + Send + Sync + Unpin + 'static,
+    ) -> anyhow::Result<Result<(), String>> {
+        self.instantiate()
+            .await
+            .context("failed to instantiate actor")?
+            .call(operation, request, response)
+            .await
+    }
+
+    /// Instantiates and returns a [`GuestInstance`] if exported by the [`Instance`].
+    ///
+    /// # Errors
+    ///
+    /// Fails if either instantiation fails or no guest bindings are exported by the [`Instance`]
+    pub async fn as_guest(&self) -> anyhow::Result<GuestInstance> {
+        self.instantiate()
+            .await
+            .context("failed to instantiate actor")?
+            .into_guest()
+            .await
+    }
+
+    /// Instantiates and returns a [`IncomingHttpInstance`] if exported by the [`Instance`].
+    ///
+    /// # Errors
+    ///
+    /// Fails if either instantiation fails or no incoming HTTP bindings are exported by the [`Instance`]
+    pub async fn as_incoming_http(&self) -> anyhow::Result<IncomingHttpInstance> {
+        self.instantiate()
+            .await
+            .context("failed to instantiate actor")?
+            .into_incoming_http()
+            .await
+    }
+
+    /// Instantiates and returns a [`LoggingInstance`] if exported by the [`Instance`].
+    ///
+    /// # Errors
+    ///
+    /// Fails if either instantiation fails or no logging bindings are exported by the [`Instance`]
+    pub async fn as_logging(&self) -> anyhow::Result<LoggingInstance> {
+        self.instantiate()
+            .await
+            .context("failed to instantiate actor")?
+            .into_logging()
+            .await
+    }
+}
+
+/// A pre-loaded, configured wasmCloud actor instance, which is either a module or a component
+#[derive(Debug)]
+pub enum Instance {
+    /// WebAssembly module containing an actor
+    Module(ModuleInstance),
+    /// WebAssembly component containing an actor
+    Component(ComponentInstance),
+}
+
+/// A pre-loaded, configured guest instance, which is either a module or a component
+#[derive(Clone)]
+pub enum GuestInstance {
+    /// WebAssembly module containing an actor
+    Module(ModuleGuestInstance),
+    /// WebAssembly component containing an actor
+    Component(ComponentGuestInstance),
+}
+
+/// A pre-loaded, configured [Logging] instance, which is either a module or a component
+pub enum LoggingInstance {
+    /// WebAssembly module containing an actor
+    Module(ModuleGuestInstance),
+    /// WebAssembly component containing an actor
+    Component(ComponentInterfaceInstance<component::logging_bindings::Logging>),
+}
+
+/// A pre-loaded, configured [`IncomingHttp`] instance, which is either a module or a component
+pub enum IncomingHttpInstance {
```

Comment:
What are the differences between `Instance`, `GuestInstance`, `LoggingInstance`, and `IncomingHttpInstance`?

Reply:
The latter 3 have exact interface bindings instantiated